Add GSoC 2023 midterm recap blog post (#6555)
jenkins-infra/jenkins.io: content/blog/2023/07/22/2023-07-22-gsoc-2023-midterm.adoc

---
layout: post
title: Jenkins in Google Summer of Code Midterm Recap
tags:
- gsoc
- gsoc2023
- midterm
authors:
- ash-sxn
- harsh-ps-2003
- jagruti
- vandit1604
opengraph:
  image: /images/gsoc/opengraph.png
sig:
- gsoc
---

The Jenkins in GSoC program recently reached its midterm milestone for GSoC 2023.
This milestone served to assess performance, and precipitate pass/fail decisions.
Taking time to evaluate the progress and workflow of the project provides an opportunity to correct course and address underlying issues.

Earlier this month, on the link:https://www.meetup.com/jenkins-online-meetup/[Jenkins Online Meetup], GSoC contributors presented midterm status and progress for their projects, below is a recap.
Thank you Jagruti, Harsh, Vandit, and Ashutosh for your contributions to Jenkins!

== link:https://www.jenkins.io/projects/gsoc/2023/projects/add-probes-to-plugin-health-score/[Add Probes to "Plugin Health Score"]
By link:https://github.com/Jagrutiti/[Jagruti Tiwari]

> A probe collects data about each plugin.
> It is an automated system ensuring that the Jenkins Community defined rules are followed.
> Probe data is used by health scoring to compute a score.
> A higher quality probe is given a good score.
> An obsolete or outdated plugin is given a lower score.

Reviewer: ```suggestion
A probe collects data about each plugin.
It is an automated system that ensures the Jenkins Community-defined rules are followed.
The probe data is used by health scoring to compute a score, and a higher-quality probe receives a good score. Conversely, an obsolete or outdated plugin is given a lower score.
```

Sharing details about the probes that I worked on during the first phase of GSoC.

=== Unreleased Production Changes Probe

> A bug fix, security fix, or new features are of no use unless they are released.
> The "unreleased production changes" probe looks for unreleased commits in the production files.
> 
> The probe aims to identify unreleased production commits in the Maven production files.
> 
> My mentors suggested I should start with writing the test cases to reproduce the probe requirements, and later add code to pass each of the test cases.
> 
> I implemented the JGit library and understood its features.
> I lacked depth in Git concepts that made me struggle with JGit.
> This led me to strengthen the Git concepts along the way.

Reviewer: ```suggestion
A bug fix, security improvement, or new feature can only be beneficial when it's actually released.
The "Unreleased Production Changes" probe is designed to detect unreleased commits in the production files.

This probe's primary goal is to identify unreleased production commits specifically in the Maven production files.

Following my mentors' advice, I began by writing test cases to accurately reproduce the probe's requirements. Subsequently, I added the necessary code to pass each of these test cases.

During the implementation, I utilized the JGit library and familiarized myself with its features.
However, I encountered some challenges due to a lack of depth in Git concepts, which required me to delve deeper and strengthen my understanding of Git concepts along the way.
```

=== The Third-Party Repository Detection Probe

> By definition, a third-party repository is a repository that is not hosted at link:https://repo.jenkins-ci.org/[].
> 
> The third-party repository detection probe ensures that plugins are built only from trusted and reliable Maven repositories.
> 
> This probe was requested by the community.
> 
> A major challenge in implementing this probe was considering the parent and child pom relationship from all angles.
> 
> Some edge-case scenarios were not considered while specifying the probe.
> And, due to the number of open questions, this probe was put on hold until further research was completed.
> 
> In the process, I have extensively read the Maven documentation to understand how to test POM structure, inheritance, and Maven API.

Reviewer: ```suggestion
By definition, a third-party repository is a repository that is not hosted at link:https://repo.jenkins-ci.org/[https://repo.jenkins-ci.org].

The third-party repository detection probe ensures that plugins are only built from trusted and reliable Maven repositories.

This probe was requested by the Jenkins community.

One of the major challenges in implementing this probe was considering the parent and child pom relationships from all angles.

Certain edge-case scenarios were not initially taken into account while specifying the probe, and due to the number of open questions, the probe was temporarily put on hold until further research was conducted.

Throughout the process, I extensively read the Maven documentation to gain a better understanding of how to test POM structure, inheritance, and Maven API.
```

> 

Reviewer 2: Read the link:blog/2023/07/20/third-party-repository-detection-probe.adoc[unreleased production changes] blog post for more details on the probe.

=== GitHub Security Scan Workflow Probe

> The Security Scan probe identifies whether the plugin is configured to execute the security action.
> 
> This probe is important from the security perspective, as it looks for known security breaches and vulnerabilities in a plugin.
> 
> Its primary purpose is to verify that the security scan is properly configured in the plugin's GitHub Action.
> 
> This probe is also involved in the refactoring of existing classes to avoid code duplication.
> The JEP-229 (Continuous  Delivery) probe looks for CI configurations in GitHub Actions as well.
> 
> Thus, I have gained knowledge in leveraging the class hierarchy to build an abstract probe that can be implemented by any probes requiring GitHub Action configuration checks.
> 
> This probe will be beneficial to the security team as it helps to identify any gaps and helps to improve the security of the Jenkins plugin ecosystem.

Reviewer: ```suggestion
The Security Scan probe identifies whether the plugin is configured to execute the security action.

From a security perspective, this probe is crucial as it checks for known security breaches and vulnerabilities in a plugin.

Its primary purpose is to ensure that the security scan is correctly configured in the plugin's GitHub Action.

Additionally, this probe involves refactoring existing classes to avoid code duplication, and it shares similarities with the JEP-229 (Continuous Delivery) probe, which also looks for CI configurations in GitHub Actions.

Throughout this process, I have gained knowledge in leveraging the class hierarchy to build an abstract probe that can be implemented by any probes requiring GitHub Action configuration checks.

The findings of this probe will be highly beneficial to the security team, as it helps identify any gaps and enables the improvement of the overall security of the Jenkins plugin ecosystem.
```

=== What about the second-half of GSoC?

> In the second half of the GSoC I will work on a few more probes like the following ones:
> 
> * Detect renovate usage
> ** Detects how many plugins use the renovate bot CLI to detect outdated dependencies in the plugin.
> ** If it finds any, it automatically creates a pull request (PR) in which the dependency is updated.
> * Count open tickets
> ** Detect open issues in Github and JIRA
> 
> Also will work on fixing bugs and making enhancements to the existing probes.

Reviewer: ```suggestion
In the second half of the GSoC project, I will be working on a few more probes, including the following:

* Detect Renovate usage:
** This probe will identify how many plugins use the Renovate bot CLI to detect outdated dependencies in the plugin.
** If any outdated dependencies are found, it will automatically create a pull request (PR) to update the dependency.
* Count open tickets:
** This probe will detect open issues in both Github and JIRA.

In addition to these new probes, I will also be focusing on fixing bugs and making enhancements to the existing probes.
```

== link:https://www.jenkins.io/projects/gsoc/2023/projects/gitlab-plugin-modernization/[GitLab Plugin Modernization]

By link:https://github.com/harsh-ps-2003/[Harsh Pratap Singh]

> The GitLab Plugin enables seamless interaction between Jenkins and GitLab.
> This project replaces the usage in the plugin of the very old RESTEasy library with GitLab4J-API via the GitLab API Jenkins Library Plugin.
> 
> This migration is important as:
> 
> 1. The migration will greatly reduce future maintenance issues.
> 2. It makes the plugin more lightweight due to reduced dependencies.
> 3. It improves the consistency with other Jenkins plugins.
> 4. Improved documentation.

Reviewer: ```suggestion
The GitLab Plugin facilitates seamless interaction between Jenkins and GitLab.
This project involves replacing the usage of the outdated RESTEasy library in the plugin with GitLab4J-API through the GitLab API Jenkins Library Plugin.

This migration holds significant importance for the following reasons:

1. The migration will significantly reduce future maintenance issues.
2. It makes the plugin more lightweight by reducing dependencies.
3. It improves consistency with other Jenkins plugins.
4. It enhances documentation and overall user experience.
```

> During the coding Phase 1 following things were achieved:
> 
> 1. Completed migration from RESTEasy to GitLab4J-API via GitLab-API plugin.
> 2. Interactively testing of the partially migrated Plugin.
> 
> To use the upcoming modernized GitLab Plugin you require:
> 
> 1. Minimum GitLab version - 14.0.
> 2. Minimum Jenkins version - 2.387.3.

Reviewer: ```suggestion
During Coding Phase 1, the following achievements were made:

1. Successfully completed the migration from RESTEasy to GitLab4J-API via the GitLab-API plugin.
2. Conducted interactive testing of the partially migrated Plugin.

To use the upcoming modernized GitLab Plugin, please ensure the following requirements are met:

1. Minimum GitLab version: `14.0`.
2. Minimum Jenkins version: `2.387.3`.
```

> Over the years GitLab has evolved significantly, thus during the migration we have found out that the plugin will have Removal of GitLab API V3 support soon as GitLab has dropped V3 and only supports V4 of its API currently. Thus, we encourage V3 users to switch to V4 as soon as possible.
> 
> Following are the things still in progress during the 2nd Coding Phase:
> 
> 1. Adapting the Webhooks to GitLab4J-API’s events (completed)
> 2. Migrating the Proxy settings
> 3. Adapting the unit and integration tests (WIP)
> 4. Improving the documentation

Reviewer: ```suggestion
Over the years, GitLab has evolved significantly.
During the migration process, we discovered that the plugin will soon have the removal of GitLab API V3 support, as GitLab now only supports V4 of its API.
We strongly encourage V3 users to switch to V4 as soon as possible.

The following tasks are still in progress during the 2nd Coding Phase:

1. Adapting the Webhooks to GitLab4J-API's events (completed)
2. Migrating the Proxy settings
3. Adapting the unit and integration tests (work in progress)
4. Improving the documentation
```

== link:https://www.jenkins.io/projects/gsoc/2023/projects/alternative-jenkinsio-build-tool/[Building Jenkins.io with Alternative Tools]
By link:https://www.jenkins.io/blog/authors/vandit1604/[Vandit Singh]

> The Jenkins users frequently refer to link:https://www.jenkins.io/[jenkins.io] for documentation, and a new LTS line is released and available to all Jenkins users worldwide every 12 weeks.
> The documentation for each prior edition is not currently maintained and that’s what my project solves by providing Versioned Documentation for the users of the Jenkins community.
> 
> Depending on the version of Jenkins they are running, my project will let the user switch between different versions of the documentation.
> We do this by adopting Antora, which comes with versioning out-of-the-box and creates documentation pages using asciidoc.
> Gatsby will be used to generate some of the YAML file-generated components.

Reviewer: ```suggestion
Jenkins users frequently refer to link:/[jenkins.io] for documentation, and a new LTS line is released and available to all Jenkins users worldwide every 12 weeks.
However, the documentation for each prior edition is not currently maintained, and that's where my project comes in to provide Versioned Documentation for the users of the Jenkins community.

With my project, users will be able to switch between different versions of the documentation, depending on the version of Jenkins they are running.
To achieve this, we adopt Antora, which comes with versioning out-of-the-box and creates documentation pages using asciidoc.
Additionally, Gatsby will be used to generate some of the YAML file-generated components.
This way, we aim to enhance the accessibility and usability of the documentation for Jenkins users.
```

> Till now, we have migrated all documentation from link:https://www.jenkins.io/[jenkins.io] which leaves us only with the Gatsby part of the implementation. The completed components include:

Reviewer: ```suggestion
Till now, we have migrated all documentation from link:https://www.jenkins.io/[jenkins.io] which leaves us only with the Gatsby part of the implementation.
The completed components include:
```

1. User Guide
2. Solution Pages
3. Tutorials
4. Developer Guide
5. Security
6. Subprojects
7. Community Section

> With the migration of the above components, we are to work on the following components/single pages using Gatsby as the framework:

Reviewer: ```suggestion
With the successful migration of the above components, our next focus will be on working with the following components or single pages using Gatsby as the framework:
```

1. Blogs
2. Roadmap
3. Security Advisories
4. Download
5. Changelog
6. Upgrade Guide
7. About
8. Homepage

> As the 2nd Coding Phase of GSoC, completing these sections/pages will mark the end of the project and the complete migration from `awestruct` to Antora and Gatsby.

Reviewer: ```suggestion
As we enter the 2nd Coding Phase of GSoC, completing these sections/pages will mark the end of the project and the full migration from awestruct to Antora and Gatsby.
```

== link:https://www.jenkins.io/projects/gsoc/2023/projects/docker-compose-build/[Docker-based Jenkins Quickstart Examples]
By link:https://www.jenkins.io/blog/authors/ash-sxn/[Ashutosh Saxena]

> The project's goal is to address the problems and complexities users often face when setting up Jenkins with Docker.
> The current process involves multiple intimidating steps with security risks.
> How are we Solving these problems?

Reviewer: ```suggestion
The project's goal is to address the problems and complexities users often face when setting up Jenkins with Docker.
The current process involves multiple intimidating steps with security risks.
So, how are we solving these problems?
```

> The intimidating steps can be made simple by using Docker Compose to hide all the complexity of Docker.
> And, we have created scripts to automate the running of the jenkins container with docker using a script called  `jenkins_init.sh` and another script cleaning everything called `jenkins_teardown.sh`.
> 
> To address the security risk of running jobs on controller, we are using the controller and agent as separate containers, another script `keygen.sh` is used that creates and updates ssh keys.

Reviewer: ```suggestion
The intimidating steps can be simplified by using Docker Compose to hide all the complexity of Docker.
Additionally, we have created scripts to automate the running of the Jenkins container with Docker.
One script is called `jenkins_init.sh`, and another script called `jenkins_teardown.sh` is used to clean everything.

To address the security risk of running jobs on the controller, we are using the controller and agent as separate containers.
For this purpose, we use another script called `keygen.sh` that creates and updates `SSH` keys.
```

> For the tutorial part, it is important that we make the barrier to entry for Jenkins as low as possible.
> So the user will need to add a keyword to the `jenkins_init.sh` script, for example for a maven tutorial the user will only need to run `./jenkins_init.sh maven` to start the tutorial.
> 
> And everything now works with `Gitpod` with just a click.

Reviewer: ```suggestion
For the tutorial part, it is essential to lower the barrier to entry for Jenkins.
Therefore, the user will only need to add a keyword to the `jenkins_init.sh` script.
For example, for a Maven tutorial, the user can simply run `./jenkins_init.sh maven` to start the tutorial.

Additionally, everything now works with `Gitpod` with just a click.
```

> Things to do during the 2nd Coding Phase:
> 
> * Adding More Tutorials and integrating them with `./jenkins_init.sh`.
> * Adding support for windows without WSL.
> * ALl the tutorial files and scripts will be tested regularly.
> * Developing concise and easy to understand documentation for new and updated tutorials.

Reviewer: ```suggestion
Things to do during the 2nd Coding Phase:

* Add more tutorials and integrate them with ./jenkins_init.sh.
* Provide support for Windows without WSL.
* Regularly test all the tutorial files and scripts.
* Develop concise and easy-to-understand documentation for new and updated tutorials.
```

During the implementation, I utilized the JGit library and familiarized myself with its features.
However, I encountered some challenges due to a lack of depth in Git concepts, which required me to delve deeper and strengthen my understanding of Git concepts along the way.

=== The Third-Party Repository Detection Probe

By definition, a third-party repository is a repository that is not hosted at link:https://repo.jenkins-ci.org/[https://repo.jenkins-ci.org].

The third-party repository detection probe ensures that plugins are only built from trusted and reliable Maven repositories.

This probe was requested by the Jenkins community.

One of the major challenges in implementing this probe was considering the parent and child pom relationships from all angles.

Certain edge-case scenarios were not initially taken into account while specifying the probe, and due to the number of open questions, the probe was temporarily put on hold until further research was conducted.

Throughout the process, I extensively read the Maven documentation to gain a better understanding of how to test POM structure, inheritance, and Maven API.

> 

Author: ```suggestion

Read the link:blog/2023/07/20/third-party-repository-detection-probe.adoc[unreleased production changes] blog post for more details on the probe.

```

By link:https://github.com/Jagrutiti/[Jagruti Tiwari]

A probe collects data about each plugin.
It is an automated system that ensures the Jenkins Community-defined rules are followed.
The probe data is used by health scoring to compute a score, and a higher-quality probe receives a good score.
Conversely, an obsolete or outdated plugin is given a lower score.

Sharing details about the probes that I worked on during the first phase of GSoC.

=== Unreleased Production Changes Probe

A bug fix, security improvement, or new feature can only be beneficial when it's actually released.
The "Unreleased Production Changes" probe is designed to detect unreleased commits in the production files.

This probe's primary goal is to identify unreleased production commits specifically in the Maven production files.

Following my mentors' advice, I began by writing test cases to accurately reproduce the probe's requirements. Subsequently, I added the necessary code to pass each of these test cases.

During the implementation, I utilized the JGit library and familiarized myself with its features.

> However, I encountered some challenges due to a lack of depth in Git concepts, which required me to delve deeper and strengthen my understanding of Git concepts along the way.

Reviewer 2: Read the link:blog/2023/07/16/unreleased-production-changes-probe.adoc[unreleased production changes] blog post for more details on the probe.

Author: Updated

Read the link:blog/2023/07/20/third-party-repository-detection-probe.adoc[unreleased production changes] blog post for more details on the probe.

=== GitHub Security Scan Workflow Probe

The Security Scan probe identifies whether the plugin is configured to execute the security action.

From a security perspective, this probe is crucial as it checks for known security breaches and vulnerabilities in a plugin.

Its primary purpose is to ensure that the security scan is correctly configured in the plugin's GitHub Action.

Additionally, this probe involves refactoring existing classes to avoid code duplication, and it shares similarities with the JEP-229 (Continuous Delivery) probe, which also looks for CI configurations in GitHub Actions.

Throughout this process, I have gained knowledge in leveraging the class hierarchy to build an abstract probe that can be implemented by any probes requiring GitHub Action configuration checks.

The findings of this probe will be highly beneficial to the security team, as it helps identify any gaps and enables the improvement of the overall security of the Jenkins plugin ecosystem.

> 

Reviewer 2: Read the link:blog/2023/07/16/security-scan-probe-blog/.adoc[security scan probe] blog post for more details on the probe.

Author: Updated

Till now, we have migrated all documentation from link:https://www.jenkins.io/[jenkins.io] which leaves us only with the Gatsby part of the implementation.
The completed components include:

1. User Guide
2. Solution Pages
3. Tutorials
4. Developer Guide
5. Security
6. Subprojects
7. Community Section

With the successful migration of the above components, our next focus will be on working with the following components or single pages using Gatsby as the framework:

1. Blogs
2. Roadmap
3. Security Advisories
4. Download
5. Changelog
6. Upgrade Guide
7. About
8. Homepage

As we enter the 2nd Coding Phase of GSoC, completing these sections/pages will mark the end of the project and the full migration from awestruct to Antora and Gatsby.

== link:https://www.jenkins.io/projects/gsoc/2023/projects/docker-compose-build/[Docker-based Jenkins Quickstart Examples]
By link:https://www.jenkins.io/blog/authors/ash-sxn/[Ashutosh Saxena]

The project's goal is to address the problems and complexities users often face when setting up Jenkins with Docker.
The current process involves multiple intimidating steps with security risks.
So, how are we solving these problems?

> The intimidating steps can be simplified by using Docker Compose to hide all the complexity of Docker.

Reviewer 3: ```suggestion
The intimidating steps can be simplified by using Docker Compose to hide all the complexities of Docker.
```
